testtools/matchers/_warnings.py

```diff
@@ -0,0 +1,97 @@
+# Copyright (c) 2009-2016 testtools developers. See LICENSE for details.
+
+__all__ = [
+    'Warnings',
+    'warning_message',
+    'is_deprecated']
+
+import warnings
+
+from ._basic import Is
+from ._const import Always
+from ._datastructures import MatchesListwise, MatchesStructure
+from ._higherorder import (
+    AfterPreprocessing,
+    Annotate,
+    MatchesAll,
+    Not,
+    )
+from ._impl import Mismatch
+
+
+def warning_message(category_type, message=None, filename=None, lineno=None,
+                    line=None):
+    """
+    Create a matcher that will match `warnings.WarningMessage`\s.
+
+    :param category_type: A warning type.
+    :param message_matcher: Match the warning message against this.
+    :param filename_matcher: Match the warning filename against this.
+    :param lineno_matcher: Match the warning line number against this.
+    :param line_matcher: Match the warning line of code against this.
+    """
+    category_matcher = Is(category_type)
+    message_matcher = message or Always()
+    filename_matcher = filename or Always()
+    lineno_matcher = lineno or Always()
+    line_matcher = line or Always()
+    return MatchesStructure(
+        category=Annotate(
+            "Warning's category type does not match",
+            category_matcher),
+        message=Annotate(
+            "Warning's message does not match",
+            AfterPreprocessing(str, message_matcher)),
+        filename=Annotate(
+            "Warning's filname does not match",
+            filename_matcher),
+        lineno=Annotate(
+            "Warning's line number does not match",
+            lineno_matcher),
+        line=Annotate(
+            "Warning's source line does not match",
+            line_matcher))
+
+
+class Warnings(object):
+    """
+    Match if the matchee produces deprecation warnings.
+    """
+    def __init__(self, warnings_matcher=None):
+        """
+        Create a Warnings matcher.
+
+        :param warnings_matcher: Optional validator for the warnings emitted by
+        matchee. If no warnings_matcher is supplied then the simple fact that
+        at least one warning is emitted is considered enough to match on.
+        """
+        self.warnings_matcher = warnings_matcher
+
+    def match(self, matchee):
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter('always')
+            matchee()
+            if not w:
```

Comment:
So, just from reading the code, I don't think it will, since you're always checking `if not w`, which will always return a Mismatch. Further, `self.warnings_matcher` is only evaluated if `if not w` evaluates to false. 

If we want to support this use case (and I don't see any reason not to), I think it may be worth adding a test case for it. I am, after all, just reading the code without actually running it (I'm feeling lazy today).